### ISSUE TITLE
KLDivLoss backward computation error

### ISSUE THREAD

**Issue by shevious** [pytorch/pytorch#89558]
### 🐛 Describe the bug

From the pytorch version 1.13.0, KLDivLoss backward computation produces `nan` gradient.
The code runs without error in the pytorch version 1.12.

```py
import numpy as np
import torch
import torch.nn as nn

torch.autograd.set_detect_anomaly(True)

p1_np = np.array([0., -700.])
p2_np = np.array([1.,    0.])
p1 = torch.from_numpy(p1_np).requires_grad_(True)
p2 = torch.from_numpy(p2_np).requires_grad_(True)

kl_loss = nn.KLDivLoss(reduction="batchmean")
output = kl_loss(p1, p2)
print(output)

output.backward()
```
The output error:
```
tensor(0., dtype=torch.float64, grad_fn=<DivBackward0>)
/home/username/venv-torch1.13/lib/python3.8/site-packages/torch/autograd/__init__.py:197: UserWarning: Error detected in XlogyBackward0. Traceback of forward call that caused the error:
  File "test.py", line 13, in <module>
    output = kl_loss(p1, p2)
  File "/home/username/venv-torch1.13/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/username/venv-torch1.13/lib/python3.8/site-packages/torch/nn/modules/loss.py", line 471, in forward
    return F.kl_div(input, target, reduction=self.reduction, log_target=self.log_target)
  File "/home/username/venv-torch1.13/lib/python3.8/site-packages/torch/nn/functional.py", line 2928, in kl_div
    reduced = torch.kl_div(input, target, reduction_enum, log_target=log_target)
  File "/home/username/venv-torch1.13/lib/python3.8/site-packages/torch/fx/traceback.py", line 57, in format_stack
    return traceback.format_stack()
 (Triggered internally at ../torch/csrc/autograd/python_anomaly_mode.cpp:114.)
  Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
Traceback (most recent call last):
  File "test.py", line 16, in <module>
    output.backward()
  File "/home/username/venv-torch1.13/lib/python3.8/site-packages/torch/_tensor.py", line 487, in backward
    torch.autograd.backward(
  File "/home/username/venv-torch1.13/lib/python3.8/site-packages/torch/autograd/__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: Function 'XlogyBackward0' returned nan values in its 1th output.
```

### Versions

Collecting environment information...
PyTorch version: 1.13.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.8.13 (default, Aug 11 2022, 09:24:24)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-25-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090

Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn.so.8.4.1
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.4.1
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.4.1
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.4.1
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.4.1
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.4.1
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.13.0+cu116
[conda] Could not collect

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

**Comment by lezcano** [comment]
The first input has a `-700`. That is not a valid input, as the inputs are the logarithms of a probability distribution. As such, the result `nan` is correct. Even more, I would argue that the fact that the forward function returns `0` in this case is rather sus, and a better implementation would be to return nan as well.

**Comment by jiayisunx** [comment]
hi @lezcano, it looks like this PR(https://github.com/pytorch/pytorch/pull/80334) is changing the behavior of KLDivLoss, is this change expected?

**Comment by ezyang** [comment]
if you have a separate issue about kldivloss please file an issue about it!

**Comment by shevious** [comment]
> The first input has a `-700`. That is not a valid input, as the inputs are the logarithms of a probability distribution. As such, the result `nan` is correct. Even more, I would argue that the fact that the forward function returns `0` in this case is rather sus, and a better implementation would be to return nan as well.

The log probability 0, -700 is occurred at the trasnfromers when the masks turns off the data with -infty and followed by log_softmax.
For example:
```python
import torch
torch.autograd.set_detect_anomaly(True)
input_data = torch.tensor([[1.5, 0.0]], requires_grad=True) 
target_prob = torch.tensor([[1.0, 0.0]], requires_grad=True)
input_logprob = torch.nn.functional.log_softmax(input_data, dim=1)
print(input_logprob)
kl_loss = torch.nn.KLDivLoss(reduction="batchmean")
output = kl_loss(input_logprob, target_prob)
output.backward()
```
The code also produces nan gradients. 
I think this situation is not so rare.

**Comment by lezcano** [comment]
So, here, you are trying to compute the derivative of `kl_div` at the point (p1, p2), (1, 0). The issue is that the function uses `torch.xlogy` in the forward applied to the point `(0,0)`. The function `f(x) = xlog(x)` is not differentiable at zero, so that's why you are getting this error: You are trying to compute the gradient of a function exactly at a point at which it is not differentiable.

A marginally better result would be to return `+inf` in this case, but I think that it would be also of little use, so IMO I think that returning `inf` is alright.

**Comment by shevious** [comment]
Hi, @lezcano,

Thanks for the clear explanation. Now I can understand the behavior of `kl_div`.  
The `kl_div` function is implemented using `torch.xlogy`, and `xlogy` has somewhat complex gradient issue at `(0,0)`. 

[https://github.com/pytorch/pytorch/issues/80770#issuecomment-1211339189](url)

The actual derivative of `f(x)=xlog(x)` at zero is `-inf`. I think it is difficult to produce correct `-inf` with `torch.xlogy` implementation of `f(x)=xlog(x)` because `xlogy` has two variables.
(The torch version 1.12 works, but returns **zero** derivative which is **incorrect** anyway.)

This problem can be avoided by clamping or adding with epsilon to the target probability of `kl_div`:
```python
import torch
eps = 1e-8
input_logprob = torch.tensor([[-0.6, -0.8]], requires_grad=True) 
#input_logprob = torch.tensor([[0.0, -700.]], requires_grad=True) 
target_prob = torch.tensor([[1.0, 0.0]], requires_grad=True)
kl_loss = torch.nn.KLDivLoss(reduction="batchmean")
#output = kl_loss(input_logprob, torch.clamp(target_prob, min=eps)) # clampping with epsilon
output = kl_loss(input_logprob, target_prob+eps) # adding with epsilon
output.backward()
print('input gradient = ', input_logprob.grad)
print('target gradient = ', target_prob.grad)

Output:
input gradient =  tensor([[-1.0000e+00, -1.0000e-08]])
target gradient =  tensor([[  1.6000, -16.6207]])
```

In the Tensorflow, the similar clipping is used in `kl_divergence`.
[https://github.com/keras-team/keras/blob/e6784e4302c7b8cd116b74a784f4b78d60e83c26/keras/losses.py#L2399-L2400](url)

I think it would be better if the `xlogy` returns correct `-inf` target gradient at `(0,0)` instead of `nan`, thus I will post new `xlogy` issue if it is needed.

**Comment by lezcano** [comment]
Just for reference, `xlogy` as a function of two variables is not differentiable at (0,0) (the derivatives approaching zero along different curves do not give the same value). If you approach (0,0) along the curve f(t) = (t, t), that is, you consider the function `xlog(x)` then that gradient happens to be -inf, and thi is our situation.

To fix this properly, we would need to add an `xlogx` with the correct derivative. But yeah, not worth it just to go from a nan to a -inf.

**Comment by shevious** [comment]
Hi, @lezcano,

> To fix this properly, we would need to add an `xlogx` with the correct derivative. But yeah, not worth it just to go from a nan to a -inf.

Thanks for the comments. I completely agree with you. I thought about `xlogx` function, too. :)

**Comment by eanson023** [comment]
```python
  def forward(self, net_outs, gts):
      mask = gts["motion_masks"]
      dn_s_logits = dn_e_logits = None
      # Find dn_loc only once per iteration
      .....
      start_logits = net_outs["grounding_start_loc"]
      end_logits = net_outs["grounding_end_loc"]
      masked_start_logits = start_logits * mask
      masked_end_logits = end_logits * mask
      masked_dn_s_logits = dn_s_logits * mask
      masked_dn_e_logits = dn_e_logits * mask
      dn_s_prob = F.log_softmax(masked_dn_s_logits, dim=-1)
      s_prob = F.softmax(masked_start_logits, dim=-1)
      start_loss = F.kl_div(dn_s_prob.detach(), s_prob, reduction='batchmean')
      # # start_loss = KL / dn_s_prob.shape[0]

      dn_e_prob = F.log_softmax(masked_dn_e_logits, dim=-1)
      e_prob = F.softmax(masked_end_logits, dim=-1)
      end_loss = F.kl_div(dn_e_prob.detach(), e_prob, reduction='batchmean')
      # end_loss = KL / dn_e_prob.shape[0]
      
      return self.kl_lambda*(start_loss + end_loss)
```

```bash
/home/k/miniconda3/envs/ys-teslam/lib/python3.7/site-packages/torch/autograd/__init__.py:199: UserWarning: Error detected in XlogyBackward0. Traceback of forward call that caused the error:
```

I got same error, but after I removed the `detach()`, the entire process of the model did not throw an exception. Adding it, an error will be reported after a certain training time.
